File: src/ladder.rs
//! Stage 1: graded text compression ladder (port of pxpipe mcp/compress.mjs).
//! Levels 0-4, each ⊇ the previous. From level 2 up, any line that looks like
//! code/data or carries a hash/path/URL/long id is passed VERBATIM.

use regex::Regex;
use std::sync::LazyLock;

pub const LEVELS: [(&str, &str, &str); 5] = [
    ("none", "none", "passthrough (baseline)"),
    (
        "whitespace",
        "lossless",
        "trailing whitespace + blank-line runs collapsed; safe for code",
    ),
    (
        "prose",
        "light",
        "L1 + prose lines: collapse spaces, cut redundant filler phrases (code/IDs protected)",
    ),
    (
        "dense",
        "medium",
        "L2 + prose: drop articles & intensifiers",
    ),
    (
        "caveman",
        "heavy",
        "L3 + prose: telegraphic — drop function words; gist only, NOT verbatim",
    ),
];

static HEDGES: LazyLock<Vec<(Regex, &'static str)>> = LazyLock::new(|| {
    [
        (r"(?i)\bit (might|may|could) (potentially )?be worth (considering|noting|mentioning) that\b", ""),
        (r"(?i)\bit is worth (noting|mentioning) that\b", ""),
        (r"(?i)\b(please )?keep in mind that\b", ""),
        (r"(?i)\bit goes without saying that\b", ""),
        (r"(?i)\bneedless to say,?", ""),
        (r"(?i)\bin my (personal )?opinion,?", ""),
        (r"(?i)\bas far as i (can tell|know),?", ""),
        (r"(?i)\bat the end of the day,?", ""),
        (r"(?i)\bmake sure to\b", ""),
        (r"(?i)\byou (should|need to|must) (make sure|ensure) that\b", "ensure"),
        (r"(?i)\bi('d| would) recommend (using|trying)\b", "use"),
        (r"(?i)\bi('d| would) recommend\b", ""),
        (r"(?i)\bto be honest,?", ""),
        (r"(?i)\bit turns out( that)?\b", ""),
    ]
    .iter()
    .map(|(p, r)| (Regex::new(p).unwrap(), *r))
    .collect()
});

static FILLER: LazyLock<Vec<(Regex, &'static str)>> = LazyLock::new(|| {
    [
        (r"(?i)\bin order to\b", "to"),
        (r"(?i)\bdue to the fact that\b", "because"),
        (r"(?i)\bat this point in time\b", "now"),
        (r"(?i)\bin the event that\b", "if"),
        (r"(?i)\bfor the purpose of\b", "for"),
        (r"(?i)\bwith regard to\b", "about"),
        (r"(?i)\ba large number of\b", "many"),
        (r"(?i)\bit is important to note that\b", ""),
        (r"(?i)\bplease note that\b", ""),
        (r"(?i)\bas a matter of fact\b", ""),
        (r"(?i)\bin terms of\b", "for"),
        (r"(?i)\bthe fact that\b", "that"),
    ]
    .iter()
    .map(|(p, r)| (Regex::new(p).unwrap(), *r))
    .collect()
});
static ARTICLES: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)\b(the|an|a)\s+").unwrap());
static INTENSIFIERS: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"(?i)\b(very|really|just|actually|basically|simply|quite|rather|essentially|literally)\s+",
    )
    .unwrap()
});
static FUNCTION_WORDS: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)\b(is|are|was|were|am|be|been|being|do|does|did|have|has|had|will|would|shall|should|can|could|may|might|of|to|in|on|at|for|with|that|this|these|those|it|its|there|here)\b\s*").unwrap()
});
static SPACES: LazyLock<Regex> = LazyLock::new(|| Regex::new(r" {2,}").unwrap());
static PUNCT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\s+([.,;:!?])").unwrap());
static NL3: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\n{3,}").unwrap());

/// A line that must be preserved verbatim: indented (code), symbol-dense
/// (code/JSON), or carrying a long whitespace-free token (hash/path/URL/id).
pub fn is_protected_line(line: &str) -> bool {
    if line.starts_with([' ', '\t']) {
        return true;
    }
    let total = line.chars().count();
    if total == 0 {
        return false;
    }
    let sym = line
        .chars()
        .filter(|c| {
            !(c.is_whitespace()
                || c.is_ascii_alphanumeric()
                || matches!(
                    c,
                    '.' | ',' | ';' | ':' | '\'' | '"' | '!' | '?' | '(' | ')' | '-'
                ))
        })
        .count();
    if sym as f64 / total as f64 > 0.3 {
        return true;
    }
    line.split_whitespace().any(|t| t.chars().count() >= 24)
}

fn tighten_prose(line: &str, level: u8) -> String {
    let mut s = SPACES.replace_all(line, " ").into_owned();
    if level >= 2 {
        for (re, to) in HEDGES.iter() {
            s = re.replace_all(&s, *to).into_owned();
        }
    }
    for (re, to) in FILLER.iter() {
        s = re.replace_all(&s, *to).into_owned();
    }
    if level >= 3 {
        s = ARTICLES.replace_all(&s, "").into_owned();
        s = INTENSIFIERS.replace_all(&s, "").into_owned();
    }
    if level >= 4 {
        s = FUNCTION_WORDS.replace_all(&s, "").into_owned();
    }
    s = SPACES.replace_all(&s, " ").into_owned();
    s = PUNCT.replace_all(&s, "$1").trim().to_string();
    // re-capitalize sentence start
    let mut chars = s.chars();
    match chars.next() {
        Some(c) if c.is_ascii_lowercase() => c.to_ascii_uppercase().to_string() + chars.as_str(),
        _ => s,
    }
}

pub struct Compressed {
    pub compressed: String,
    pub protected_lines: usize,
    pub level: u8,
}

pub fn compress_text(text: &str, level: u8) -> Compressed {
    let level = level.min(4);
    if level == 0 {
        return Compressed {
            compressed: text.to_string(),
            protected_lines: 0,
            level,
        };
    }
    let mut protected = 0usize;
    let lines: Vec<String> = text
        .split('\n')
        .map(|raw| {
            let line = raw.trim_end_matches([' ', '\t']);
            if level == 1 {
                return line.to_string();
            }
            if is_protected_line(line) {
                protected += 1;
                return line.to_string();
            }
            tighten_prose(line, level)
        })
        .collect();
    let joined = lines.join("\n");
    Compressed {
        compressed: NL3.replace_all(&joined, "\n\n").into_owned(),
        protected_lines: protected,
        level,
    }
}

#[cfg(test)]
mod hedge_tests {
    use super::*;

    #[test]
    fn test_hedges_at_level_2() {
        let input = "It might be worth noting that this works.";
        let c = compress_text(input, 2);
        assert_eq!(c.compressed, "This works.");
    }

    #[test]
    fn test_hedges_ignored_at_level_1() {
        let input = "It might be worth noting that this works.";
        let c = compress_text(input, 1);
        assert!(c.compressed.contains("might be worth"));
    }

    #[test]
    fn test_h10_replacement() {
        let input = "You should make sure that the file exists.";
        let c = compress_text(input, 2);
        assert_eq!(c.compressed, "Ensure the file exists.");
    }

    #[test]
    fn test_h11_replacement() {
        let input = "I would recommend using this approach.";
        let c = compress_text(input, 2);
        assert_eq!(c.compressed, "Use this approach.");
    }

    #[test]
    fn test_protected_line_with_hedge() {
        let input = "    it might be worth noting that code = true;";
        let c = compress_text(input, 2);
        // Protected lines bypass tighten_prose entirely
        assert!(c.compressed.contains("might be worth"));
        assert_eq!(c.protected_lines, 1);
    }

    #[test]
    fn test_multiple_hedges() {
        let input = "Needless to say, in my opinion, this works.";
        let c = compress_text(input, 2);
        assert_eq!(c.compressed, "This works.");
    }

    #[test]
    fn test_hedge_with_optional_comma() {
        let input = "To be honest this is fine.";
        let c2 = compress_text(input, 2);
        assert_eq!(c2.compressed, "This is fine.");

        let input2 = "To be honest, this is fine.";
        let c3 = compress_text(input2, 2);
        assert_eq!(c3.compressed, "This is fine.");
    }
}

File: src/ladder.ts
// Stage 1: graded text compression ladder (port of pxpipe mcp/compress.mjs).
// Levels 0-4, each ⊇ the previous. From level 2 up, any line that looks like
// code/data or carries a hash/path/URL/long id is passed VERBATIM.

import { isRustWhitespace, rustTrim } from "./serde.ts";

export const LEVELS: [string, string, string][] = [
  ["none", "none", "passthrough (baseline)"],
  [
    "whitespace",
    "lossless",
    "trailing whitespace + blank-line runs collapsed; safe for code",
  ],
  [
    "prose",
    "light",
    "L1 + prose lines: collapse spaces, cut redundant filler phrases (code/IDs protected)",
  ],
  [
    "dense",
    "medium",
    "L2 + prose: drop articles & intensifiers",
  ],
  [
    "caveman",
    "heavy",
    "L3 + prose: telegraphic — drop function words; gist only, NOT verbatim",
  ],
];

// Rust regex `\b` is Unicode-aware: \w = Alphabetic + Mark + Decimal_Number +
// Connector_Punctuation + Join_Control. JS `\b` is ASCII-only, so emulate the
// boundaries with lookarounds against the Unicode word class.
const W =
  "[\\p{Alphabetic}\\p{Mark}\\p{Decimal_Number}\\p{Connector_Punctuation}\\p{Join_Control}]";
// Rust `\s` / char::is_whitespace = Unicode White_Space. JS `\s` differs
// (misses U+0085 NEL, adds U+FEFF), so use the explicit property set.
const SP =
  "[\\t\\n\\x0B\\f\\r \\u0085\\u00A0\\u1680\\u2000-\\u200A\\u2028\\u2029\\u202F\\u205F\\u3000]";

// `(?i)\b<phrase>\b` — phrase starts/ends on a word char in every entry.
const wordPhrase = (p: string): RegExp =>
  new RegExp(`(?<!${W})(?:${p})(?!${W})`, "giu");

const FILLER: [RegExp, string][] = [
  [wordPhrase("in order to"), "to"],
  [wordPhrase("due to the fact that"), "because"],
  [wordPhrase("at this point in time"), "now"],
  [wordPhrase("in the event that"), "if"],
  [wordPhrase("for the purpose of"), "for"],
  [wordPhrase("with regard to"), "about"],
  [wordPhrase("a large number of"), "many"],
  [wordPhrase("it is important to note that"), ""],
  [wordPhrase("please note that"), ""],
  [wordPhrase("as a matter of fact"), ""],
  [wordPhrase("in terms of"), "for"],
  [wordPhrase("the fact that"), "that"],
];
// (?i)\b(the|an|a)\s+
const ARTICLES = new RegExp(`(?<!${W})(?:the|an|a)${SP}+`, "giu");
// (?i)\b(very|...)\s+
const INTENSIFIERS = new RegExp(
  `(?<!${W})(?:very|really|just|actually|basically|simply|quite|rather|essentially|literally)${SP}+`,
  "giu",
);
// (?i)\b(is|...)\b\s*
const FUNCTION_WORDS = new RegExp(
  `(?<!${W})(?:is|are|was|were|am|be|been|being|do|does|did|have|has|had|will|would|shall|should|can|could|may|might|of|to|in|on|at|for|with|that|this|these|those|it|its|there|here)(?!${W})${SP}*`,
  "giu",
);
const SPACES = / {2,}/g;
const PUNCT = new RegExp(`${SP}+([.,;:!?])`, "gu");
const NL3 = /\n{3,}/g;

/**
 * A line that must be preserved verbatim: indented (code), symbol-dense
 * (code/JSON), or carrying a long whitespace-free token (hash/path/URL/id).
 */
export function isProtectedLine(line: string): boolean {
  if (line.length === 0) return false;
  const c0 = line.charCodeAt(0);
  if (c0 === 0x20 || c0 === 0x09) return true;
  let total = 0;
  let sym = 0;
  let run = 0; // current whitespace-free token length in codepoints
  let longTok = false;
  for (let i = 0; i < line.length; ) {
    const cp = line.codePointAt(i)!;
    i += cp > 0xffff ? 2 : 1;
    total++;
    if (isRustWhitespace(cp)) {
      run = 0;
      continue;
    }
    run++;
    if (run >= 24) longTok = true;
    // ascii alnum
    if (
      (cp >= 0x30 && cp <= 0x39) ||
      (cp >= 0x41 && cp <= 0x5a) ||
      (cp >= 0x61 && cp <= 0x7a)
    ) {
      continue;
    }
    // . , ; : ' " ! ? ( ) -
    switch (cp) {
      case 0x2e: // .
      case 0x2c: // ,
      case 0x3b: // ;
      case 0x3a: // :
      case 0x27: // '
      case 0x22: // "
      case 0x21: // !
      case 0x3f: // ?
      case 0x28: // (
      case 0x29: // )
      case 0x2d: // -
        continue;
    }
    sym++;
  }
  if (sym / total > 0.3) return true;
  return longTok;
}

function tightenProse(line: string, level: number): string {
  let s = line.replace(SPACES, " ");
  for (let i = 0; i < FILLER.length; i++) {
    s = s.replace(FILLER[i][0], FILLER[i][1]);
  }
  if (level >= 3) {
    s = s.replace(ARTICLES, "");
    s = s.replace(INTENSIFIERS, "");
  }
  if (level >= 4) {
    s = s.replace(FUNCTION_WORDS, "");
  }
  s = s.replace(SPACES, " ");
  s = rustTrim(s.replace(PUNCT, "$1"));
  // re-capitalize sentence start (ASCII lowercase only, like Rust)
  const first = s.charCodeAt(0);
  if (first >= 0x61 && first <= 0x7a) {
    s = String.fromCharCode(first - 0x20) + s.slice(1);
  }
  return s;
}

export interface Compressed {
  compressed: string;
  protectedLines: number;
  level: number;
}

export function compressText(text: string, level: number): Compressed {
  // Clamp defensively at the choke point, matching Rust's `f64 as u8` (which
  // saturates) followed by `.min(4)`. `Math.min(level, 4)` alone let a
  // negative through: the proxy's `--level -1` reached tightenProse, which
  // only tests >=3 and >=4, so it silently applied LOSSY level-2 prose
  // compression here while Rust applied none. Every caller is now safe
  // regardless of how it parsed the number.
  const lvl = Math.min(Math.max(0, Math.trunc(level)), 4);
  if (lvl === 0) {
    return { compressed: text, protectedLines: 0, level: lvl };
  }
  let protectedLines = 0;
  const parts = text.split("\n");
  const lines: string[] = new Array(parts.length);
  for (let i = 0; i < parts.length; i++) {
    const raw = parts[i];
    // trim_end_matches([' ', '\t'])
    let end = raw.length;
    while (end > 0) {
      const c = raw.charCodeAt(end - 1);
      if (c !== 0x20 && c !== 0x09) break;
      end--;
    }
    const line = end === raw.length ? raw : raw.slice(0, end);
    if (lvl === 1) {
      lines[i] = line;
      continue;
    }
    if (isProtectedLine(line)) {
      protectedLines++;
      lines[i] = line;
      continue;
    }
    lines[i] = tightenProse(line, lvl);
  }
  const joined = lines.join("\n");
  return {
    compressed: joined.replace(NL3, "\n\n"),
    protectedLines,
    level: lvl,
  };
}
